lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/WordStorage.java

```
@@ -0,0 +1,338 @@
+   *       the difference of this entry's start and the prefix's entry start. 0 for single-character
+   *       entries
+   *   <li>Optional, for non-leaf entries only:
+   *       <ul>
+   *         <li>VINT: the length of the word form data, returned from {@link #lookupWord}
+   *         <li>n * VINT: the word form data
+   *         <li>Optional, for hash-colliding entries only:
+   *             <ul>
+   *               <li>BYTE: 1 if the next collision entry has further collisions, 0 if it's the
+   *                   last of the entries with the same hash
+   *               <li>VINT: (relative) pointer to the previous entry with the same hash
+   *             </ul>
+   *       </ul>
+   * </ul>
+   */
+  private final byte[] wordData;
+
+  private WordStorage(int[] hashTable, byte[] wordData) {
+    this.hashTable = hashTable;
+    this.wordData = wordData;
+  }
+
+  IntsRef lookupWord(char[] word, int offset, int length) {
+    assert length > 0;
+
+    int hash = Math.abs(CharsRef.stringHashCode(word, offset, length) % hashTable.length);
+    int pos = hashTable[hash];
+    if (pos == 0) {
+      return null;
+    }
+
+    boolean collision = pos < 0;
+    pos = Math.abs(pos);
+
+    char lastChar = word[offset + length - 1];
+    ByteArrayDataInput in = new ByteArrayDataInput(wordData);
+    while (true) {
+      in.setPosition(pos);
+      char c = (char) in.readVInt();
+      int prevPos = pos - in.readVInt();
+      int beforeForms = in.getPosition();
+      boolean found = c == lastChar && isSameString(word, offset, length - 1, prevPos, in);
+      if (!collision && !found) {
+        return null;
+      }
+
+      in.setPosition(beforeForms);
+      int formLength = in.readVInt();
+      if (found) {
+        IntsRef forms = new IntsRef(formLength);
+        readForms(forms, in, formLength);
+        return forms;
+      } else {
+        skipVInts(in, formLength);
+      }
+
+      collision = in.readByte() == 1;
```

Instead of burning whole byte to represent "next entry still has a collision", you could steal one bit from the `delta pos` you read next?  It'd make this a bit more compact, and would be consistent with the up front `pos < 0` collision check which is single bit encoding.

---

(`delta` is a great word, I'll use it, thanks!)

Oh, appending the lowest bit to the delta looks nice, I haven't thought of this! However, I've another idea for using that byte: to store the word length, which might speed up lookup and especially enumeration. I plan to test whether this is true, and if not, will append the bit.

---

Good idea (storing word length) -- that'd make lookups (that hashed to the same mod location but have different lengths) fail faster.